#8 fix user create bug

vadzim-matsiuk opened the pull request:
fixed #8

Commit:
fix 403 forbidden

Commit:
fixed trouble with db

mikita-herasiutsin commented:
We have a new problem. For successful POST request we receive entity with all information including password, e.g:
{
    "id": 11,
    "username": "bob3",
    "password": "secret",
    "encryptedPassword": "$2a$10$oAIb1xYOPItDhZpp/pmYpOsIF2Qh/.4FKxmEsDiTdUQlnz4GRLuNS"
}

I suggest to change response to 201 (Created) with empty body

Commit:
solved receive problem

We received entity with all informations  after the controller had processed POST. To solve the problem, ResponseEntity with created status was used.

mikita-herasiutsin commented:
Add integration tests for saving user

Commit:
added integration test

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -0,0 +1,39 @@
+   ResponseEntity<User> responseEntity = restTemplate.postForEntity("/user",
+       user, User.class);
+
+   assertEquals(HttpStatus.CREATED, responseEntity.getStatusCode());
```
Use assertThat() in all tests and add checking for empty response body

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -0,0 +1,39 @@
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT )
+public class UserControllerIT {
```
Delete another empty file for integration tests

Commit:
added checking for empty response body

Commit:
used assertThat instead of assertTrue

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -36,6 +35,6 @@ public void createUser() {
 
    assertThat(HttpStatus.CREATED, is(responseEntity.getStatusCode()));
-   assertTrue(!responseEntity.hasBody());
+   assertThat(false, is(responseEntity.hasBody()));
```
`assertThat(actual, expected)` but not` assertThat(expected, actual)`. Also `assertThat(responseEntity.getBody(), is(nullValue()))` seems more readable

Commit:
AssertThat  was fixed

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -34,7 +35,7 @@ public void createUser() {
-   assertThat(false, is(responseEntity.hasBody()));
+   assertThat(responseEntity.getStatusCode(), is(HttpStatus.CREATED));
+   assertThat(responseEntity.getBody(), is(nullValue()));
   }
```
you need to check db that user is saved

Commit:
improved integration test

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -35,6 +44,9 @@ public void createUser() {
 
+   User savedUser = userRepository.findByUsername("test");
+
+   assertThat(savedUser, is(equalTo(user)));
```
is(equalTo()) is redundant because is() calls equalTo()

mikita-herasiutsin commented on src/main/java/simplifier/model/User.java:
```diff
@@ -53,4 +55,25 @@ public String getEncryptedPassword() {
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```
The problem is user with same username can be saved and then retrieving 2 users from repository with same usernames but different id and passwords will be equal

mikita-herasiutsin commented on pom.xml:
```diff
@@ -63,7 +63,12 @@
+      <artifactId>commons-lang3</artifactId>
+      <version>3.4</version>
+    </dependency>
+  </dependencies>
```
format this file

mikita-herasiutsin replied:
Replace tabs to spaces in all files

Commit:
added check for duplicate username

Now we can't save user in the database if username doesn't unique

mikita-herasiutsin commented on src/main/java/simplifier/services/UserServiceImpl.java:
```diff
@@ -25,7 +26,11 @@ public void setbCryptPasswordEncoder(
+  public User saveUser(User user) throws NameNotUniqueException{
+    User checkForUnique = userRepository.findByUsername(user.getUsername());
+    if (checkForUnique != null) {
+      throw new NameNotUniqueException();
```
Change throwing exception to `Optional<User>` return type

mikita-herasiutsin commented on src/main/java/simplifier/controllers/UserController.java:
```diff
@@ -23,9 +24,11 @@ public void setUserService(UserService userService) {
-    userService.saveUser(user);
-
+    try {
+      userService.saveUser(user);
```
Remove try/catch. This type of design should be avoided

Commit:
used Optional instead of throwing exception

mikita-herasiutsin commented on src/main/java/simplifier/services/UserServiceImpl.java:
```diff
@@ -26,12 +26,12 @@ public void setbCryptPasswordEncoder(
   @Override
-  public User saveUser(User user) throws NameNotUniqueException{
+  public Optional<User> saveUser(User user) {
     User checkForUnique = userRepository.findByUsername(user.getUsername());
```
Variable names should be noun but not verb. Change variable name to just user or foundUser

mikita-herasiutsin commented on src/test/java/simplifier/UserServiceImplTest.java:
```diff
@@ -36,12 +34,7 @@ public void saveUser() {
 
     when(userRepository.save(user)).thenReturn(user);
 
```
All cases should be covered by unit tests

mikita-herasiutsin commented on src/main/java/simplifier/controllers/UserController.java:
```diff
@@ -24,11 +23,9 @@ public void setUserService(UserService userService) {
-      userService.saveUser(user);
-    } catch (NameNotUniqueException e) {
-      return ResponseEntity.status(HttpStatus.CONFLICT).body("Username already exists");
+    if (userService.saveUser(user).isPresent()) {
```
Use map()/orElse() methods instead of if()/isPresent()

Commit:
added unit & integration tests

Also project was formatted

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -16,37 +12,59 @@
-  }
+    @Test
+    public void conflictIfUsernameNotUnique() {
+        User firstUser = new User();
```
Use more meaningful names for variables, e.g. existingUser and another invalidUser

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -16,37 +12,59 @@
-   assertThat(responseEntity.getStatusCode(), is(HttpStatus.CREATED));
-   assertThat(responseEntity.getBody(), is(nullValue()));
-  }
+        assertThat(responseEntity.getStatusCode(), is(HttpStatus.CONFLICT));
```
Verifying that only one user is saved and response body has message are required

mikita-herasiutsin commented on src/test/java/simplifier/UserServiceImplTest.java:
```diff
@@ -16,28 +10,54 @@
-  private UserServiceImpl userService;
+    @Test
+    public void doNotSaveIfUsernameDuplicate() {
+        User firstUser = new User();
```
Use more meaningful names

mikita-herasiutsin commented on src/test/java/simplifier/UserServiceImplTest.java:
```diff
@@ -16,28 +10,54 @@
 
-  @Mock
-  private BCryptPasswordEncoder bCryptPasswordEncoder;
+        verify(userRepository).save(user);
```
You should verify that password was encoded

mikita-herasiutsin commented on src/test/java/simplifier/UserServiceImplTest.java:
```diff
@@ -16,28 +10,54 @@
 
-    verify(userRepository).save(user);
-  }
+        verify(userRepository).findByUsername(secondUser.getUsername());
```
You should verify that password encoding hasn't been invoked

Commit:
fixed tests

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -0,0 +1,73 @@
+
+        assertThat(responseEntity.getStatusCode(), is(HttpStatus.CONFLICT));
+        assertThat(responseEntity.getBody(), is(notNullValue()));
+        assertThat(userRepository.findByUsername("test"), is(existingUser));
```
You verify that found user is existing one but verification uses user's equals() method which compares usernames only. It means there can be 2 users with same usernames. This verification does not do what it should. I suggest to change equals()  to compare all fields

mikita-herasiutsin commented on src/test/java/simplifier/UserServiceImplTest.java:
```diff
@@ -17,28 +10,57 @@
 
-    verify(userRepository).save(user);
-  }
+        verify(userRepository).findByUsername(invalidUser.getUsername());
```
Add verification that no more invocations of this mock are executed. It is possible to use save() method for example and test will pass, same is for bCryptPasswordEncoder. Use verifyNoMoreInteractions() in all tests or use should().then() instead of verify()

mikita-herasiutsin commented on src/test/java/simplifier/UserControllerIT.java:
```diff
@@ -0,0 +1,73 @@
+                invalidUser, String.class);
+
+        assertThat(responseEntity.getStatusCode(), is(HttpStatus.CONFLICT));
+        assertThat(responseEntity.getBody(), is(notNullValue()));
```
Verify that exactly message exists in body

mikita-herasiutsin commented on src/test/java/simplifier/UserServiceImplTest.java:
```diff
@@ -1,13 +1,6 @@
 package simplifier;
```
Move all tests to same packages that tests classes are in

Commit:
Tests were improved

mikita-herasiutsin commented on src/test/java/simplifier/controllers/UserControllerIT.java:
```diff
@@ -57,7 +67,7 @@ public void conflictIfUsernameNotUnique() {
         existingUser.setPassword("test");
 
-        userRepository.save(existingUser);
+        Optional<User> savedUser = userService.saveUser(existingUser);
```
Services are not used in tests. Use repository instead

mikita-herasiutsin commented on src/test/java/simplifier/controllers/UserControllerIT.java:
```diff
@@ -46,7 +56,7 @@ public void createUser() {
         User savedUser = userRepository.findByUsername("username");
 
-        assertThat(savedUser, is(user));
+        assertThat(savedUser, is(notNullValue()));
```
Why did you change to noNullValue()?

mikita-herasiutsin commented on src/test/java/simplifier/controllers/UserControllerIT.java:
```diff
@@ -0,0 +1,86 @@
+        String requiredMessage = "Username already exists";
+
+        assertThat(responseEntity.getStatusCode(), is(HttpStatus.CONFLICT));
+        assertThat(responseEntity.getBody(), is(notNullValue()));
```
this is redundant line

Commit:
integration tests were fixed
